### PR TITLE
Add basic PlantUML support (feature request #78)

### PR DESCRIPTION
**Pull request by studix**
- ArchUnitNET.Domain.PlantUml: 
new namespace which contains the PlantUML parser code ported from ArchUnit java

- TypesShould / TypeConditionsDefinition:
contains new method  _AdhereToPlantUmlDiagram_ for PlantUML based dependency checks

ExampleArchUnitTestPuml demonstrates the feature

### CODE REVIEW

**Comment by fgather** [comment]
something bad happened :) The diff contains all Commits since some years, which makes it very hard to review. Do you know how this happened? Looks like you recommitted all the history, only with the original author?

File: ArchUnitNET/Domain/PlantUml/ComponentName.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+        public override bool Equals(object obj)
+        {
+            return obj is ComponentName name &&
+                   _value == name._value;
+        }
+
+        public override int GetHashCode()
+        {
+            return -1939223833 + EqualityComparer<string>.Default.GetHashCode(_value);
```
Comment (fgathercoupa) [inline ArchUnitNET/Domain/PlantUml/ComponentName.cs:28]: how did you generate the equals/gethashcode methods? would it be possible to use the convention we have in other places?

File: ArchUnitNET/Domain/PlantUml/PlantUmlDiagram.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Runtime.CompilerServices;
+
+[assembly: InternalsVisibleTo("ArchUnitNETTests")]
+namespace ArchUnitNET.Domain.PlantUml
+{
+    internal class PlantUmlDiagram
```
Comment (fgathercoupa) [inline ArchUnitNET/Domain/PlantUml/PlantUmlDiagram.cs:9]: should all those domain classes be internal? or could some be accessible to be used without the fluent api? Like most other domain classes

File: ArchUnitNET/Domain/PlantUml/PlantUmlParser.cs [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+using System.Text;
+using System.Text.RegularExpressions;
+using static ArchUnitNET.Domain.PlantUml.PlantUmlPatterns;
+
+namespace ArchUnitNET.Domain.PlantUml
+{
+    internal class PlantUmlParser
+    {
+        private PlantUmlPatterns _plantUmlPatterns = new PlantUmlPatterns();
+        public PlantUmlDiagram Parse(string filename)
```
Comment (fgathercoupa) [inline ArchUnitNET/Domain/PlantUml/PlantUmlParser.cs:16]: we could additionally allow for streams. That could make it easier to test against architecture documentations, but more importantly: we could simplify some test code, which would not need to write files, but the test files could be inserted as an in memory stream

File: ArchUnitNETTests/Domain/PlantUml/TestDiagram.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        {
+            _path = path;
+        }
+
+        public static TestDiagram In(string path)
+        {
+            return new TestDiagram(path);
+        }
+
+        public ComponentCreator Component(string componentName)
```
Comment (fgathercoupa) [inline ArchUnitNETTests/Domain/PlantUml/TestDiagram.cs:25]: suggestion: rename to ComponentBuilder? That might improve consistency with other builders in the project

File: ExampleTest/Resources/shopping_example.puml [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+[Customers] --> [Orders]
+
+[Products] <--[#green]- catalog
+[Products] --> [Customers]
+[Products] --> [Orders]
+catalog --> [Orders]
+import -left-> catalog : parse products
+import --> xml
+import --> [Customers]
+note top on link #lightgreen: is responsible for translating XML to java classes
```
Comment (fgathercoupa) [inline ExampleTest/Resources/shopping_example.puml:36]: Java? What is this :D

**Review by fgathercoupa — CHANGES_REQUESTED** [review]
Thank you! The code looks very clean to me and I only have a few remarks :)

**Comment by studix** [comment]
fgathercoupa> something bad happened :) The diff contains all Commits since some years, which makes it very hard to review. Do you know how this happened? Looks like you recommitted all the history, only with the original author?

I initially forgot to add a signoff message to my commits. In the pull request a rebase command was suggested by github to fix this. Unfortunatly this command also added my signoff message to old commits. I would suggest to reject my pull request. I'm worried about that this will be merged also to the main branch when you accept the pull request. I could take over my changes to a new branch in my fork and create a new pull request. What do you think?

File: ArchUnitNET/Domain/PlantUml/ComponentName.cs (hunk 1 above)
Comment (studix) [reply ArchUnitNET/Domain/PlantUml/ComponentName.cs:28]: Visual Studio 2019 generated that code. I will have a look at other equals methods and use that convention.

File: ArchUnitNET/Domain/PlantUml/PlantUmlDiagram.cs (hunk 2 above)
Comment (studix) [reply ArchUnitNET/Domain/PlantUml/PlantUmlDiagram.cs:9]: My intention was to encapsulate the whole parser code. This code should not by part of the libraries public API as it is an implementation detail.

File: ArchUnitNET/Domain/PlantUml/PlantUmlParser.cs (hunk 3 above)
Comment (studix) [reply ArchUnitNET/Domain/PlantUml/PlantUmlParser.cs:16]: Good idea.

File: ExampleTest/Resources/shopping_example.puml (hunk 5 above)
Comment (studix) [reply ExampleTest/Resources/shopping_example.puml:36]: copy & paste

File: ArchUnitNET/Domain/PlantUml/PlantUmlDiagram.cs (hunk 2 above)
Comment (fgathercoupa) [reply ArchUnitNET/Domain/PlantUml/PlantUmlDiagram.cs:9]: ok, sounds reasonable. However, this somehow implies that this feature won't be able with the non-fluent API. which is fine, but maybe there is a nice way to enable it later.

File: ArchUnitNET/Domain/PlantUml/ComponentName.cs (hunk 1 above)
Comment (fgathercoupa) [reply ArchUnitNET/Domain/PlantUml/ComponentName.cs:28]: I can also do this. I assume we used Rider/Resharper

**Comment by fgathercoupa** [comment]
> 
> 
> fgathercoupa> something bad happened :) The diff contains all Commits since some years, which makes it very hard to review. Do you know how this happened? Looks like you recommitted all the history, only with the original author?
> 
> I initially forgot to add a signoff message to my commits. In the pull request a rebase command was suggested by github to fix this. Unfortunatly this command also added my signoff message to old commits. I would suggest to reject my pull request. I'm worried about that this will be merged also to the main branch when you accept the pull request. I could take over my changes to a new branch in my fork and create a new pull request. What do you think?

Hi, 
I won't merge it accidentally, but yes we should fix it. Maybe it is possible to cherry-pick your commits into a clean branch? And change the source branch to the new branch?

**Comment by studix** [comment]
> > fgathercoupa> something bad happened :) The diff contains all Commits since some years, which makes it very hard to review. Do you know how this happened? Looks like you recommitted all the history, only with the original author?
> > I initially forgot to add a signoff message to my commits. In the pull request a rebase command was suggested by github to fix this. Unfortunatly this command also added my signoff message to old commits. I would suggest to reject my pull request. I'm worried about that this will be merged also to the main branch when you accept the pull request. I could take over my changes to a new branch in my fork and create a new pull request. What do you think?
> 
> Hi,
> I won't merge it accidentally, but yes we should fix it. Maybe it is possible to cherry-pick your commits into a clean branch? And change the source branch to the new branch?

What do you mean with "change the source branch to the new branch"?

**Comment by studix** [comment]
I would have a clean new branch ready on my fork. It also contains the changes from the review. Shall we delete this pull request and I will create a new one?

**Comment by fgather** [comment]
Hi,

yes, that would be great, thank you!

Best Regards

**Comment by studix** [comment]
Pull request replaced by pull request https://github.com/TNG/ArchUnitNET/pull/86

**Pull request closed without merging**